File: solutions/ex2_allgather.f90
program allgather
use mpi_f08
integer, parameter :: n=64
integer i,iloc,proc,nproc, B(n)
integer, allocatable :: A(:)
call MPI_Init()
call MPI_COMM_RANK(MPI_COMM_WORLD, proc)
call MPI_COMM_SIZE(MPI_COMM_WORLD, nproc)
if (proc==0 .and. mod(n,nproc) /= 0) then
   print *, 'Number of processes incompatible with the size'
   call MPI_Abort(MPI_COMM_WORLD,1)
endif
allocate (A(n/nproc))
iloc=1
do i=proc*n/nproc+1, (proc+1)*n/nproc
   a(iloc)=i**2
   iloc=iloc+1
enddo
call MPI_Allgather(A,n/nproc,MPI_INTEGER, &
                   B,n/nproc,MPI_INTEGER,MPI_COMM_WORLD)
print *, 'proc ',proc,' A = ',A
print *, 'proc ',proc,' B = ',B
deallocate (A)
call MPI_Finalize()
end program allgather

File: solutions/ex2_allgather.c
#include <stdio.h>
#include <mpi.h>
#include <stdlib.h>
#define n 64

int main(int argc, char *argv[])
{

int i,iloc,ierr,proc,nproc, B[n];
int *A;

ierr = MPI_Init(&argc,&argv);

ierr = MPI_Comm_size(MPI_COMM_WORLD, &nproc);
ierr = MPI_Comm_rank(MPI_COMM_WORLD, &proc);

if ((proc==0) && (n%nproc != 0)) {
   printf("Number of processes incompatible with the size\n");
   ierr = MPI_Abort(MPI_COMM_WORLD,1);
}
A = (int *) malloc((n/nproc)*sizeof(int));
iloc=0;
for ( i=proc*n/nproc; i<(proc+1)*n/nproc ; i++) {
   A[iloc]=(i+1)*(i+1);
   iloc++;
}
ierr = MPI_Allgather(A,n/nproc,MPI_INT,
                   B,n/nproc,MPI_INT,MPI_COMM_WORLD);

printf("proc %i A = ", proc);
for (i = 0; i < n/nproc; i++)
   printf("%i ", A[i]);
printf("\n");

printf("proc %i B = ", proc);
for (i = 0; i < n; i++)
   printf("%i ", B[i]);
printf("\n");

free(A);
ierr = MPI_Finalize();

return 0;
}
